### ISSUE TITLE
Some questions about loss implementation

### ISSUE THREAD

**Issue by kyuyeonpooh** [hche11/Localizing-Visual-Sounds-the-Hard-Way#2]
Hi, thank you for sharing your awesome code.

I'm facing some issues while understanding your code.
 
![111](https://user-images.githubusercontent.com/22637557/118638680-0ce21f80-b812-11eb-83df-ae0922406338.PNG)
https://github.com/hche11/Localizing-Visual-Sounds-the-Hard-Way/blob/509acf8e673332e7db52ca1a7f29d65f19ef7c86/model.py#L54-L79

Referring to the code and your paper, 
I understood that `sim1` in L#69 is representing *P_i*, and `sim2` in L#72 is implementing the left term of *N_i*.

It seems `sim` in L#71 is referring to the right term of *N_i*.
However, I cannot understand why `Pos_all` variable is a thresholded value of `A0`.
I thought it should be all-one matrices, according to the paper.
- **Could you clarify where `sim` belongs to in the loss objective?**

One more question, please.
- **Regarding L#74-77, should I scale the logits with the temperature value (0.07) ?**
I am little confused as scaling logits with temperature value is not directly stated in the paper.

It would be very helpful if you can release the code of your loss function.
Thank you very much.
Have a nice day!

**Comment by hche11** [comment]
Hi, thank you for your interests and feedback. 

- **`sim`** refers to easy negatives generated from image 1 and other audios 2 - n. It can be generated by either max pooling from the heatmap or using weighted mean of the heatmap similar to **`sim1`**. We use the latter way, i.e.,`Pos_all` is the threshold which refers to the most attentive region for easy negative pairs.  I will release the loss function soon.

- Thank you. we will update the temperature value in the paper.

**Comment by jzuern** [comment]
Hi, thanks for your response.
I am trying to apply your approach on a similar dataset and was wondering about the actual loss objective function. Following the equations in your paper and the explanations above, I arrived at the following implementation:

```
heatmap, logits, Pos, Neg = model(img, spec, args)

# Sum up over spatial dimensions
Pos = torch.sum(Pos, dim=(1, 2))
Neg = torch.sum(Neg, dim=(1, 2))

# Optimization objective
bsize = Pos.shape[0]
nc_loss = -1/bsize * torch.sum(torch.log(torch.exp(Pos) / (torch.exp(Pos) + torch.exp(Neg))))

```

Can you confirm this?

Thanks
